### DISCUSSION TITLE
Getting multiple records from a supabase function

### DISCUSSION THREAD

**Discussion by RobertBrunhage** [supabase/supabase#4245] [Questions]
I have the following function that returns a `record` but this only seems to return the **first** record, instead of all of them.

![image](https://user-images.githubusercontent.com/26344867/144398717-9649af14-f50e-457d-ba01-217ad4b913ce.png)

I've tried the same query in the SQL editor/tester and during that it returns multiple records.

How do I make the function return all the records that satisfy the query?

**Comment by keybits** [comment]
I'm wondering if this is a limitation of the Supabase functions editor.

I believe you need to choose `SETOF RECORD` as described here: https://dba.stackexchange.com/a/186262

But I don't see how to do that from the Supabase editor as the type isn't listed during function creation:

![image](https://user-images.githubusercontent.com/809823/144439453-a00a4c99-4319-45e7-8b6f-716598084eaf.png)

Looking into this I also discovered that a number of options aren't available when editing a function but which are available when creating a new one (e.g. `return type`). Perhaps this will be addressed in a future update?

**Reply by RobertBrunhage** [reply]
I created an issue out of this, as that seems to make sense and should be an option if that is the case!

**Reply by keybits** [reply]
👍 Good idea and [confirmed on the issue](https://github.com/supabase/supabase/issues/4248#issuecomment-985119446) that the function UI doesn't currently support returning a record set

**Reply by schweppes-0x** [reply]
creating the function manualy (with code) helped me, use the SET OF record

**Reply by mikekok** [reply]
This still hasn't been added to the functions editor as of today.

**Reply by MildTomato** [reply]
cc @joshenlim @alaister

**Reply by richmeij** [reply]
bumping this. Still not available.

**Reply by dorskiegithub** [reply]
bumping this, badly needed the procedure for complex queries that is hard or maybe impossible to do using the query building.

**Comment by Chloe-Meinshausen** [comment]
really missing this functionality.

**Comment by horatiu665** [comment]
I would also like to get multiple records from a complex SQL query, but I didn't manage to use SQL to create a function with my custom SQL in it. I got as far as creating the function inside the SQL editor with code from https://dba.stackexchange.com/a/186262,  and the result looks like it returns SETOF record as seen in the screenshot, however it didn't work when calling the function.
<img width="978" alt="Screenshot 2024-01-26 at 13 11 22" src="https://github.com/supabase/supabase/assets/5824753/1c04bd52-72f6-46aa-9fab-0e858f0d9d0b">

However, for my use case I managed to solve it using a [view](https://www.restack.io/docs/supabase-knowledge-supabase-create-view-guide). Unfortunately, creating a view also requires custom SQL and the supabase UI does not give feedback that it created a view elsewhere in the dashboard, but it works and here is how I did it:

Once I had my query, I used the SQL editor to run:
```
CREATE VIEW latest_positions AS
CUSTOM_QUERY
```
where CUSTOM_QUERY was my specific query involving joining tables and whatnot.

I ran this code in the SQL editor, and it returned zero rows. So far so good. Then I checked in the table editor, and lo and behold, besides my previous 3 tables, there is a new entry with a different icon, one that you cannot create using the regular UI (or at least I haven't found that option), which is a view - which behaves exactly like a table, and after testing, it returns the correct data I want. See screenshot.
<img width="309" alt="Screenshot 2024-01-26 at 13 14 24" src="https://github.com/supabase/supabase/assets/5824753/2556ebab-2e30-4f10-9e8c-963848507841">
To use the view, I use it like any other table:
```
SELECT * FROM latest_positions
```

Hope this helps someone.

**Reply by alexanderchanis** [reply]
Thanks! Do you know if there's a way to do this with a function that requires an argument from the client side?

**Reply by horatiu665** [reply]
no idea! sorry! I'm not on this project anymore!

**Comment by psyberfly** [comment]
I need this to work. 
When I run the function SQL query in SQL Editor I get multiple rows.
When I run my function from the FLutter API SDK, I only get the first row of the SQL query result.
